src/librustc/traits/fulfill.rs

```diff
@@ -99,13 +104,98 @@ pub struct PendingPredicateObligation<'tcx> {
     pub stalled_on: Vec<Ty<'tcx>>,
 }
 
+/// An obligation which cannot be fulfilled in the context
+/// it was registered in, such as auto trait obligations on
+/// `impl Trait`, which require the concrete type to be
+/// available, only guaranteed after finishing type-checking.
+#[derive(Clone, Debug)]
+pub struct DeferredObligation<'tcx> {
+    pub predicate: ty::PolyTraitPredicate<'tcx>,
+    pub cause: ObligationCause<'tcx>
+}
```

Comment:
I haven't read deeply here but I'm a bit dubious about the deferred obligations strategy. *Even if* we implement things this way, we may want to consider enforcing some kind of inference DAG on uses of `impl Trait` to give us room to consider later (I guess stabilization makes this less of a concern, actually, never mind).

Anyway, I'm thinking about situations like this one:

```rust
trait Foo {
    type X;
}

impl<T> Foo for T {
    default type X = ();
}

impl<T: Send> Foo for T {
    type X = u32;
}
```

Now if you have an obligation like `<Y as Foo>::X = u32`, where `Y` is an anonymous type, we have to figure out how to resolve it, and to do that we kind of need to know whether `Y: Send`. 

In principle, we could probably defer both obligations (and maybe you do, as I said, I didn't finish reading this commit yet) but I do wonder if there will be other situations (method dispatch?) where it is very difficult to suspend.